src/modules/companies/companies.service.ts

```diff
@@ -0,0 +1,66 @@
+import { Injectable } from '@nestjs/common';
+import { CompaniesHouseService } from '@ukef/helper-modules/companies-house/companies-house.service';
+import { GetCompanyCompaniesHouseResponse } from '@ukef/helper-modules/companies-house/dto/get-company-companies-house-response.dto';
+
+import { SectorIndustryEntity } from '../sector-industries/entities/sector-industry.entity';
+import { SectorIndustriesService } from '../sector-industries/sector-industries.service';
+import { GetCompanyResponse, Industry } from './dto/get-company-response.dto';
+
+@Injectable()
+export class CompaniesService {
+  constructor(
+    private readonly companiesHouseService: CompaniesHouseService,
+    private readonly sectorIndustriesService: SectorIndustriesService,
+  ) {}
+
+  async getCompanyByRegistrationNumber(registrationNumber: string): Promise<GetCompanyResponse> {
```

Comment:
`Can this function throw any exception, if yes can it be please be added to the function signature.`
As far as I'm aware, Typescript does not have this feature: https://www.learningtypescript.com/articles/why-typescript-doesnt-include-a-throws-keyword

Reply:
`Might be prudent to wrap it in a try-catch block.`
Good shout, I've added this. The catch block throws a `NotFoundException` (which will be converted by Nest into a `404` response) if `this.companiesHouseService.getCompanyByRegistrationNumber` throws a `CompaniesHouseNotFoundException`; otherwise, it rethrows the error (so it will be unhandled and hence Nest will convert it into a `500` response).

I think that this is the behaviour that we want because:
- if the Companies House API returns a `404` (and so our Companies House Service throws a `CompaniesHouseNotFoundException`), it is useful for us to return a `404` to DTFS, as it tells the user that the company registration number they searched for cannot be found.
- if the Companies House API returns any other error response, it is not useful (and in fact misleading) to return this to DTFS, so we map these all to a `500`:
    - `400`: we do request validation *before* we send our request to CH API, so if CH API returns a `400`, it should be due to a problem in MDM (e.g., we don't construct our `'Authorization'` header properly in the code).
    - `401`: if CH API returns a `401`, the most likely problem is an incorrect API key in MDM, so it is a `500` as far as DTFS is concerned.
    - `422`: similar to the `400`, this suggests a problem in MDM, not in the request from DTFS.
    - `429`: I think we should have our services' rate-limits set to a sensible level to avoid hitting the rate-limits of 3rd party services. If we do get a `429` from CH API, I think it's more useful to log this error in MDM (which we do anyway) so we can adjust our rate-limit and return a `500` to the user.
    - `500`: a `500` from CH API clearly isn't due to the request from DTFS, so a `500` makes sense here also.

Reply:
Understood thank you.